### PR TITLE
Widgets work with both ~users and Teams

### PR DESCRIPTION
**Pull request by mattbk**
I took a crack at aligning the default widget code with the new public.json output from Gratipay.com.  Comments are welcome; I'm sure there are ways to break what I've done.

### CODE REVIEW

**Comment by whit537** [comment]
Nice! I've made a `Review` label in `grtp.co` so we could apply it here. :)

Anyone up for reviewing this? If not I'll get to it when I can ...

**Comment by mattbk** [comment]
Should I worry about these checks failing?

**Comment by whit537** [comment]
@mattbk Yes. It looks like our linter ain't happy with your formatting. ;-)

**Comment by mattbk** [comment]
Okay, I see it now.

   lib/v1/api.js
    131 |                    text('receiving', (data.taking)? '$' + data.taking : '$' + data.receiving);
                             ^ [W015] Expected 'text' to have an indentation at 17 instead at 21.
    141 |                    }
                             ^ [W015] Expected '}' to have an indentation at 17 instead at 21.

**Comment by mattbk** [comment]
I think tests are based on the old `public.json` and are failing because I downloaded new versions of the test `public.json` files?

**Comment by mattbk** [comment]
Looking at these files in order?

> Files: test/test_custom-widget-legacy.js, 
test/test_custom-widget.js, 
test/test_default-widget-anonymous.js,
 test/test_default-widget-legacy.js,
 test/test_default-widget.js, 
test/test_expose-api-to-iframe.js, 
test/test_giving-widget-anonymous.js, 
test/test_giving-widget-legacy.js, 
test/test_giving-widget.js -> test

Means the first broken test is test/test_default-widget-anonymous.js, I think because we don't have an `anonymous` value anymore.

**Comment by mattbk** [comment]
I think it's failing on this part of `api.js` for certain tests.

```
  // Set up links and username before requesting public.json
            link('profile-link', gratipayURI + encodeURIComponent(username) + '/');
            link('link', gratipayURI);
            text('username', username);
            text('identity', 'I');
```

**Comment by mattbk** [comment]
That works. Hot.

**Comment by whit537** [comment]
/me waiting for Travis to run green ...

**Comment by mattbk** [comment]
> /me waiting for Travis to run green ...

Hold your horses, I have to understand how tests work first.

**Comment by whit537** [comment]
!m @mattbk

**Comment by mattbk** [comment]
It looks like only giving can be anonymous now, not receiving. That doesn't help my brain work better, though.

**Comment by mattbk** [comment]
Previous commit/revert tells me the issue is in api.js, not v1.js.

This must be fascinating to watch.

**Comment by whit537** [comment]
@mattbk at https://github.com/gratipay/inside.gratipay.com/issues/280#issuecomment-185306575:

> I think I've gotten the base functionality down (at least for the limited scope of the PR), but I'm not sure how to make Travis okay with it. I think the issue is looking for ~user data, but getting team data from public.json and failing.
>
Any insight would be appreciated.

**Comment by whit537** [comment]
lol

screen shot 2016-02-17 at 1 22 03 pm

**Comment by whit537** [comment]
/me figuring out how to run tests for grtp.co ...

**Comment by whit537** [comment]
`.travis.yml` doesn't explicate ...

**Comment by whit537** [comment]
> For projects using npm, Travis CI will execute
>
```
npm test
```
>
to run your test suite.

https://docs.travis-ci.com/user/languages/javascript-with-nodejs#Default-Test-Script

**Comment by whit537** [comment]
Okay, I get the same result locally that I'm seeing in the latest Travis build here on this PR.

screen shot 2016-02-17 at 1 26 58 pm

**Comment by whit537** [comment]
Let's see if tests pass on master ...

**Comment by whit537** [comment]
Yes!

screen shot 2016-02-17 at 1 27 52 pm

**Comment by whit537** [comment]
Alright, looks like all six failing tests are failing in the same [way](https://travis-ci.org/gratipay/grtp.co/builds/109691959#L405): they're expecting a `readystatus` of `ready`, but are seeing `loading api.js step 3` instead.

@mattbk Why do you think it breaks [here](https://github.com/gratipay/grtp.co/blob/efd1b3032e5634e3834748d6e50a145a1e62d589/lib/v1/api.js#L240)?

One practice that I find helpful is to isolate a single test case, such that, even though six are failing here, I only focus on one for debugging purposes. Do you know how to do that with `npm test`?

Let's think in terms of the `default-widget-anonymous: readystatus should be ready` test ...

Given the observed behavior thus far, what's your hypothesis about why that test is failing? What's a test we can perform to validate your hypothesis? Like, can we `console.log` some variables at some point to see what they are? (I'm not sure what interactive debugging facilities are available for node).

Remember, debugging is Science Lite&trade;! :microscope: :older_man:

**Comment by mattbk** [comment]
> @mattbk Why do you think it breaks [here](https://github.com/gratipay/grtp.co/blob/efd1b3032e5634e3834748d6e50a145a1e62d589/lib/v1/api.js#L240)?

I added those status updates with step 1, step 2, step 3. It gets to step 3 and then (I think) goes into `widget.setAttribute('data-gratipay-username', username);` and doesn't end up setting the status as ready. I think this is because `data-gratipay-username` is empty for some reason.

I could be way off on that. I don't know how to do much of anything relating to actual frameworks, but I can probably learn the basics of `npm test` if that will let me check code without going through the commit process each time.

**Comment by whit537** [comment]
> I think this is because username is empty for some reason.

A hypothesis!

> I don't know how to do much of anything relating to actual frameworks, but I can probably learn the basics of npm test if that will let me check code without going through the commit process each time.

Have you been editing on GitHub through their web editor? Or are you set up to edit locally? Do you have `npm` installed locally? What operating system are you running?

**Comment by mattbk** [comment]
Locally, OSX. I'll have to look for npm when I get in front of that computer, but I went through the whole install/local build in the readme.

**Comment by whit537** [comment]
Okay, cool. Me, too. Do you use brew? If not I believe `npm` comes with the package installer on [nodejs.org](https://nodejs.org/en/).

**Comment by whit537** [comment]
Yup.

screen shot 2016-02-17 at 5 15 58 pm

**Comment by mattbk** [comment]
Yes, I use brew. Will look at this fresh locally when I get a chance. Thanks for taking a look.

**Comment by whit537** [comment]
Thanks for working on it! Once the tests are passing I'll give this a proper review. I need to understand what the purpose is here and also what our widget code is already like. :)

**Comment by mattbk** [comment]
For fun while simulations are running, installed Chocolatey and then npm in Windows 7. Then had to initialize/start local grunt with `npm install` in the grtp.co directory, then successfully ran `npm test` and got similar output. So there's a more streamlined approach to testing.

I don't know what it all means, but I can make it work.

**Comment by whit537** [comment]
Ah, nice! !m @mattbk 

Check the npm docs, you should be able to run just a single test pretty easily.

**Comment by mattbk** [comment]
It's going into (or maybe not even going into) Gratipay.widget.apply() and not coming out again, so `data-gratipay-readystatus` is never set as `ready`.

**Comment by mattbk** [comment]
What a difference me accidently deleting a line makes. *facepalm* 

`widget.setAttribute('data-gratipay-username', username);` was gone, so `username` wasn't making it from one function to another. Things went faster once I realized I could use the JS console in Chrome.

**Comment by whit537** [comment]
Nice! Scientific progress goes boink! :-)

**Comment by mattbk** [comment]
I ended up merging `trailing-zero` into this branch because this is what it was based on. Next up is a full review by me of what I actually did.

**Comment by mattbk** [comment]
I'm satisfied with the final product; it's a heck of a mess of commits, though. :wolf:

**Comment by aandis** [comment]
>  it's a heck of a mess of commits

Github now allows [squash and merge](https://github.com/blog/2141-squash-your-commits) commits! :dancer:

File: lib/v1/api.js [hunk 1]
```diff
@@ -127,12 +127,20 @@
...
+                text('receiving', (data.taking)? '$' + parseFloat(data.taking).toFixed(2) : '$' + parseFloat(data.receiving).toFixed(2));
                 text('username', data.username);
                 text('my-tip-button', data.my_tip == 'self' ? 'You!' : (+data.my_tip > 0 ? data.my_tip : 'Gratipay')); // jshint ignore:line
                 text('goal', '$' + data.goal);
                 text('giving', (data.giving)? '$' + data.giving : 'anonymously');
-                link('profile-link', gratipayURI + encodeURIComponent(data.username) + '/');
+
+                // If username exists, it's a ~user, not a team. Build URL accordingly.
+                if (data.username) {
+                    link('profile-link', gratipayURI + encodeURIComponent(data.username) + '/');
```
Comment (whit537) [inline lib/v1/api.js]: Don't we want to insert a `~` here?

**Comment by whit537** [comment]
As I read it, this PR tries to treat the `data-gratipay-username` of widgets-in-the-wild as *either* a ~user `username` *or* a Team `slug`.

 1. If it's a `username`, we use that to construct the profile link, and expect that `receiving` won't be present but `taking` will be.
 1. If it's a `slug`, we use *that* to construct the profile link, and expect `receiving` to present.

I'm not sure that's the right approach to take, both because it introduces tech debt into our widget code ("receiving" in this PR no longer means "receiving," it means "receiving *or* taking"), and because it's not as true to our new data model as we could be. I think a better approach will be to let `data-gratipay-username` continue to only mean ~user `username`, and write new widget code for Teams. See https://github.com/gratipay/grtp.co/issues/103#issuecomment-207886406 for further discussion.

I propose that we whittle this PR down to just the test changes that make sense, and pick up with additional changes in new PRs.

File: lib/v1/api.js (hunk 1 above)
Comment (mattbk) [inline lib/v1/api.js]: I'm not sure how those rewrites end up working. I know that if I don't put a ~ in, it still works. If it is better to include the ~, then we can do that.

**Comment by mattbk** [comment]
> I propose that we whittle this PR down to just the test changes that make sense to keep, and pick up with additional changes in new PRs.

I can work on that, unless someone else wants to jump in.

**Comment by mattbk** [comment]
>  just the test changes

I might need clarification. You mean make this PR about ~users and build a new PR for teams?

**Comment by whit537** [comment]
I actually meant to keep the extra info in the test strings like below, and make new PRs for ~user and Team changes.

screen shot 2016-04-20 at 11 27 06 am

**Comment by mattbk** [comment]
Got it. Now I get to learn how to clone my PR without losing things.

**Comment by mattbk** [comment]
Shuttering this pull request because it changes v1 widgets, which is not the plan at https://github.com/gratipay/grtp.co/issues/103#issuecomment-215227273.

Keeping the branch in place until I grab what I want for a real v2.

**Pull request closed without merging**